### ISSUE TITLE
Disconnect with Initial Injection

### ISSUE THREAD

**Issue by huitema** [quicwg/base-drafts#1951]
Suppose that a connection has progressed and the server has sent an Initial packet with a Server Hello to the client. At that point, we expect that all following packets will be protected by the handshake key or by the application key, preventing 3rd parties from interfering with the connection. 

Suppose now that a third party sends to the server a spoofed initial packet, containing for example a malformed acknowledgement. The server will process the packet, find the malformed frame, generate a protocol violation error and close the connection. Which means that the 3rd party is able to interfere with the connection after the handshake keys have already been exchanged.

Is that an attack we should worry about?

**Comment by ianswett** [comment]
Fixing this basically requires implicit ack of Initial packets, which I expect all implementations to do anyway.  I'll try to create a PR.

**Comment by ianswett** [comment]
Martin points out that HRR makes this rather odd, because the client can't ACK the HRR, so the server has to send both the HRR and ServerHello in the second flight.  Which is ok, but ugly.

**Comment by huitema** [comment]
I think there are two issues that are interacting. We want to limit what the server sends without some continuity test, to prevent amplification attacks. And then we want to prevent injection of initial packets after the handshake key is set, in order to prevent DOS on the connection. The HRR/ServerHello issue is tied to amplification attack prevention.

**Comment by siyengar** [comment]
I think we shouldn’t do this until we resolve the congestion controller exemptions for this type of vulnerable data. 

I’m not sure whether that is finished yet. Given that I would be fine with this

**Comment by kazuho** [comment]
FWIW, I am starting to wonder if there could be an alternative approach than dropping Initial keys.

Specifically, I am wondering if just having the rules for dealing with potentially attacking frames (instead of packets) would be either equivalent to (or even better than) trying to drop the keys.

To me it seems that declaring the following three rules (quoted from https://github.com/quicwg/base-drafts/pull/2053#discussion_r237723449) is sufficient, or even better considering invalid ACK injection attacks during the exchange of Initial packets. The other benefit is that the loss recovery / congestion control logic can remain unchanged.

> How about stating that, for Initial packets:
>
> * an endpoint MUST ignore CRYPTO frames once it obtains Handshake keys
> * an endpoint MUST ignore CONNECTION_CLOSE frames once it receives a Handshake packet
> * an endpoint MUST ignore unexpected packet numbers found in an ACK frame

**Comment by marten-seemann** [comment]
That seems like a very fragile solution to me. There are a lot more error conditions that than you describe (e.g. frames that are invalid in an Initial packet, incomplete frames, unknown frame types etc.).

It also feels wrong to me to act on some frames, and ignore other frames sent in the same packet. Either we have reason to believe that a packet was sent by an attacker, then we should drop it on the floor as a whole, or we believe it was sent by the peer, then we should process it entirely.

Dropping the keys as soon as possible has the beautiful property that we don’t have to worry about complex rules which packets are valid are which ones aren’t any more - since we know for sure that either way we don’t need to read these packets.

**Comment by kazuho** [comment]
I agree that dropping at per-frame basis is more fragile than dropping keys. However, it has the following benefits:
* The rule to evade attacks before the Handshake keys are exchanged. Specifically, ignoring unexpected PNs in ACK is a nice-to-have for #2076.
* The rules are easier to implement than what we have now in #2053, because the rules can be applied in one-path, rather than in two-pass (i.e. first check that the frames are valid then apply all of them).

To put it another way, per-frame rules is something easy to implement, and also is beneficial to evade attacks during the exchange of Initials. And if we use the framework also for ignoring CRYTO frames once the Handshake key is obtained, dropping the Initial keys become unnecessary.

**Comment by marten-seemann** [comment]
> The rules are easier to implement than what we have now in #2053, because the rules can be applied in one-path, rather than in two-pass (i.e. first check that the frames are valid then apply all of them).

I don't think that's true. Consider the case where a packet contains two ACK frames, the first of which is valid, the second one acknowledges an unsent packet. When processing the frames, you'd first happily process (and clean up your history of sent packets), only to encounter the second, invalid frame. What do you do now? You *could* unwind the effects of the first ACK frame, which at least in my implementation would a huge amount of complexity, or you could ignore the second ACK frame, accepting the fact that you acted upon a packet that was probably sent by an attacker.

**Comment by ianswett** [comment]
In my experience, per-frame rules are very problematic.  Particularly any that drop some frames and process others without closing the connection.

That's a complex road I'm unwilling to go down.

**Comment by kazuho** [comment]
@marten-seemann In this specific case, that is not an issue.

Before the Handshake keys are exchanged, you need to rely on ACKs. At that moment, having rules to reject certain ACKs never hurts.
After the Handshake keys are known, endpoints can ignore ACK frames entirely.

So two-pass is not required if you drop frames at per-frame level.

**Comment by marten-seemann** [comment]
> After the Handshake keys are known, endpoints can ignore ACK frames entirely.

Which requires implicit acking, along the lines of #2045, which I think is the best fix we have for the injection attack.

> Before the Handshake keys are exchanged, you need to rely on ACKs. At that moment, having rules to reject certain ACKs never hurts.

True, but it doesn't fix this issue. There are various ways to defend against injections during the 3-way handshake, and implementations may do something fancy here, but I don't think we have the tools to completely prevent this attack yet.

**Comment by kazuho** [comment]
>> After the Handshake keys are known, endpoints can ignore ACK frames entirely.
> 
> Which requires implicit acking, along the lines of #2045

It does not require implicit ACKing, because ACKs are always sent.

**Comment by marten-seemann** [comment]
If we ignore ACK frames for Initial packets after handshake keys become available, why would we still send them? And if we don't ignore them, we're vulnerable to injection attacks.

**Comment by kazuho** [comment]
> If we ignore ACK frames for Initial packets after handshake keys become available, why would we still send them?

I am saying that ignoring suspicious ACK (either partially or entirely) does not disrupt a handshake once an endpoint has obtained the handshake keys, because the effect is the same as using implicit ACKs.

The benefit of exchanging ACKs (that can be ignored) rather than doing implicit ACKs is that they provide valuable information to loss recovery and congestion control.

>  And if we don't ignore them, we're vulnerable to injection attacks.

I disagree. Assuming that we adopt the rules in 
https://github.com/quicwg/base-drafts/issues/1951#issuecomment-443065734, I think that we would be as safe as #2054 after handshake keys are obtained. Also, the proposed approach is less vulnerable to injection attacks before the handshake keys are obtained.

**Comment by mikkelfj** [comment]
@marten-seemann  wrote:

> I'm not sure if this is true. Consider the case where a packet contains two ACK frames, the first of which is valid, the second one acknowledges an unsent packet. When processing the frames, you'd first happily process (and clean up your history of sent packets), only to encounter the second, invalid frame. What do you do now? You could unwind the effects of the first ACK frame, which at least in my implementation would a huge amount of complexity, or you could ignore the second ACK frame, accepting the fact that you acted upon a packet that was probably sent by an attacker.

Whatever we do, we need to recognise that this particular case can also be an optimistic ACK attack by a valid endpoint that misbehaves. It is more likely to be a misbehaving client though. Therefore, special casing around this needs to close the connection once handshake encryption is online, but not before. That is a fair bit of complexity right there.

**Comment by martinthomson** [comment]
We should close this once #2045 is merged.  It's not perfect, but we get most of the way.

**Comment by ekr** [comment]
I want to put a marker in that I object to this change and I don't think we
had consensus, so I'm going to ask to discuss it in Tokyo.

**Comment by martinthomson** [comment]
Discussed in Tokyo:  conclusion will be added to #2267.  We're considering a solution that addresses a number of issues.
